File: 9.Algorithms: Dynamic Programming/1.fibonacci.py
def fibonacci(n):
    global calculations
    calculations += 1
    if n < 2:
        return n
    return fibonacci(n-1) + fibonacci(n-2)

calculations = 0
print(fibonacci(15)) # O(2^n) time
print(f'calc 15: {calculations}')

def fibonacci_better():
    cache = {}
    def inner_fib(n):
        global calculations2
        calculations2 += 1
        if n in cache:
            return cache[n]
        else:
            if n < 2:
                return n
            else:
                cache[n] = inner_fib(n-1) + inner_fib(n-2)
                return cache[n]
    return inner_fib

fib_dynamic = fibonacci_better()

calculations2 = 0
print(fib_dynamic(15)) # O(n) time
print(f'calc 15: {calculations2}')

def fibonacci_bottom_up(n):
    fibs = [0,1]
    for i in range(2, n+1):
        fibs.append(fibs[i-1] + fibs[i-2])
    return fibs.pop()

print(fibonacci_bottom_up(15))

File: 9.Algorithms: Dynamic Programming/1.fibonacci.js
let calculations = 0;
function fibonacci(n) {
    calculations++;
    if(n < 2) {
        return n;
    }
    return fibonacci(n-1) + fibonacci(n-2);
}

console.log(fibonacci(15)) // O(2^n) time
console.log('bad fib', calculations)

let calculations2 = 0;
function fibonacciBetter() {
    let cache = {}
    return function fib(n) {
        calculations2++;
        if(n in cache) {
            return cache[n]
        } else {
            if(n < 2) {
                return n;
            } else {
                cache[n] = fib(n-1) + fib(n-2);
                return cache[n];
            }
        }
    }
}

const fibDynamic = fibonacciBetter();
console.log(fibDynamic(15)) // O(n) time
console.log('good fib', calculations2)

function fibonacciBottomUp(n) {
    let arr = [0, 1];
    for(let i = 2; i <= n; i++) {
        arr.push(arr[i-1] + arr[i-2]);
    }
    return arr.pop();
}

console.log(fibonacciBottomUp(15))